File: jserver-security/src/main/java/com/github/gr3gdev/jserver/security/http/TokenRequest.java
package com.github.gr3gdev.jserver.security.http;

import com.github.gr3gdev.jserver.http.Request;

import java.util.Arrays;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicReference;

/**
 * TokenRequest.
 *
 * @author Gregory Tardivel
 */
public class TokenRequest {

    protected static final String AUTH = "Authorization";
    protected static final String COOKIES = "Cookie";

    /**
     * Get JWT token from Authorization header.
     */
    public static Optional<String> getTokenFromHeader(final Request req) {
        // Token in Authorization
        final AtomicReference<Optional<String>> token = new AtomicReference<>(Optional.empty());
        req.headers(AUTH).ifPresent(header -> {
            if (header.toLowerCase().startsWith("bearer")) {
                token.set(Optional.of(header.substring("Bearer ".length())));
            }
        });
        return token.get();
    }

    /**
     * Get JWT token from Cookie.
     */
    public static Optional<String> getTokenFromCookie(final Request req, final String cookieName) {
        // Token in cookie
        final AtomicReference<Optional<String>> token = new AtomicReference<>(Optional.empty());
        req.headers(COOKIES).ifPresent(ch -> {
            final String[] cookies = ch.split(" ");
            final String tokenCookie = Arrays.stream(cookies)
                    .filter(c -> c.startsWith(cookieName + "="))
                    .findFirst().orElse("");
            if (!tokenCookie.isBlank() && tokenCookie.contains("=")) {
                final String tokenValue = tokenCookie.split("=")[1];
                if (tokenValue.endsWith(";")) {
                    token.set(Optional.of(tokenValue.substring(0, tokenValue.length() - 1)));
                } else {
                    token.set(Optional.of(tokenValue));
                }

            }
        });
        return token.get();
    }

}

File: jserver-security/src/main/kotlin/com/github/gr3gdev/jserver/security/http/TokenRequest.kt
package com.github.gr3gdev.jserver.security.http

import com.github.gr3gdev.jserver.http.Request
import java.util.*


object TokenRequest {

    private const val AUTH = "Authorization"
    private const val COOKIES = "Cookie"

    /**
     * Get JWT token from Authorization header.
     */
    fun getTokenFromHeader(req: Request): Optional<String> {
        // Token in Authorization
        var token = Optional.empty<String>()
        req.headers(AUTH).ifPresent {
            if (it.startsWith("Bearer", true)) {
                token = Optional.of(it.substring("Bearer ".length))
            }
        }
        return token
    }

    /**
     * Get JWT token from Cookie.
     */
    fun getTokenFromCookie(req: Request, cookieName: String): Optional<String> {
        // Token in cookie
        var token = Optional.empty<String>()
        req.headers(COOKIES).ifPresent { ch ->
            val cookies = ch.split(" ")
            val tokenCookie = cookies.find { c -> c.startsWith("$cookieName=") }.orEmpty()
            if (tokenCookie.isNotEmpty() && tokenCookie.contains("=")) {
                val tokenValue = tokenCookie.split("=")[1]
                token = if (tokenValue.endsWith(";")) {
                    Optional.of(tokenValue.substring(0, tokenValue.length - 1))
                } else {
                    Optional.of(tokenValue)
                }

            }
        }
        return token
    }

}
